<a href="https://colab.research.google.com/github/zuzannazak/random-files-i-need-sometimes/blob/main/SID_projekt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# *Wielokrotna imputacja*

Wielokrotna imputacja to technika radzenia sobie z brakami danych w zbiorze danych i może być wykorzystywana także jako część kompletowania zbioru danych w oparciu o podzbiory pochodzące z różnych źródeł.

Wielokrotna imputacja polega na uzupełnieniu brakujących wartości rozkładem zaobserwowanych danych by uwzględnić niepewność przy estymacie prawidłowej wartości.

Implementacje wielokrotnej imputacji w R:


*   Pakiet MICE - https://cran.r-project.org/web/packages/mice/index.html
*   Pakiet Amelia - https://cran.r-project.org/web/packages/Amelia/index.html 
*   Pakiet norm - https://www.rdocumentation.org/packages/norm/versions/1.0-9.5
*   Pakiet missForest - https://cran.r-project.org/web/packages/missForest/index.html
*   Pakiet mi - https://cran.r-project.org/web/packages/mi/index.html



## Przykład 1 - dane testowe

### Przygotowanie przykładowych danych do imputacji

In [3]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


In [4]:
%%R
dat <- read.csv(url("https://goo.gl/4DYzru"), header=TRUE, sep=",")
head(dat)

   Age Gender Cholesterol SystolicBP  BMI Smoking Education
1 67.9 Female       236.4      129.8 26.4     Yes      High
2 54.8 Female       256.3      133.4 28.4      No    Medium
3 68.4   Male       198.7      158.5 24.1     Yes      High
4 67.9   Male       205.0      136.0 19.9      No       Low
5 60.9   Male       207.7      145.4 26.7      No    Medium
6 44.9 Female       222.5      130.6 30.6      No       Low


In [25]:
%%R
#Sprawdzanie wartości pustych
sapply(dat, function(x) sum(is.na(x)))
#Zamiana zmiennych na factory lub wartości liczbowe
library(dplyr) 
dat <- dat %>%
    mutate(
        Smoking = as.factor(Smoking),
        Education = as.factor(Education),
        Cholesterol = as.numeric(Cholesterol)
    )
str(dat)

'data.frame':	250 obs. of  7 variables:
 $ Age        : num  67.9 54.8 68.4 67.9 60.9 44.9 49.9 55.1 57.5 77.2 ...
 $ Gender     : chr  "Female" "Female" "Male" "Male" ...
 $ Cholesterol: num  236 256 199 205 208 ...
 $ SystolicBP : num  130 133 158 136 145 ...
 $ BMI        : num  26.4 28.4 24.1 19.9 26.7 30.6 27.3 27.5 28.3 29.1 ...
 $ Smoking    : Factor w/ 2 levels "No","Yes": 2 1 2 1 1 1 NA 1 1 1 ...
 $ Education  : Factor w/ 3 levels "High","Low","Medium": 1 3 1 2 3 2 3 2 1 1 ...


In [33]:
%%R
#Dodanie wartości pustych dla przykładu
original = dat
set.seed(10)
dat[sample(1:nrow(dat), 20), "Cholesterol"] <- NA
dat[sample(1:nrow(dat), 20), "Smoking"] <- NA
dat[sample(1:nrow(dat), 20), "Education"] <- NA
dat[sample(1:nrow(dat), 5), "Age"] <- NA
dat[sample(1:nrow(dat), 5), "BMI"] <- NA
sapply(dat, function(x) sum(is.na(x)))

        Age      Gender Cholesterol  SystolicBP         BMI     Smoking 
          5           0          20           0           5          20 
  Education 
         20 


In [27]:
%%R
#Zamiana zmiennych na factory lub wartości liczbowe
library(dplyr) 
dat <- dat %>%
    mutate(
        Smoking = as.factor(Smoking),
        Education = as.factor(Education),
        Cholesterol = as.numeric(Cholesterol)
    )
str(dat)

'data.frame':	250 obs. of  7 variables:
 $ Age        : num  67.9 54.8 68.4 67.9 60.9 44.9 49.9 55.1 57.5 77.2 ...
 $ Gender     : chr  "Female" "Female" "Male" "Male" ...
 $ Cholesterol: num  236 256 199 205 208 ...
 $ SystolicBP : num  130 133 158 136 145 ...
 $ BMI        : num  26.4 28.4 24.1 19.9 26.7 30.6 27.3 27.5 28.3 29.1 ...
 $ Smoking    : Factor w/ 2 levels "No","Yes": 2 1 2 1 1 1 NA 1 1 1 ...
 $ Education  : Factor w/ 3 levels "High","Low","Medium": 1 3 1 2 3 2 3 2 1 1 ...


### Imputacja

In [11]:
%%R
install.packages("mice")
library(mice) #Przewidywanie wartości pustych na podstawie innych zmiennych
init = mice(dat, maxit=0) 
meth = init$method
predM = init$predictorMatrix
predM[, c("BMI")]=0 #Wyłączenie zmiennej BMI z przewidywania
meth[c("Age")]="" #Ta zmienna nie będzie imputowana

R[write to console]: Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

R[write to console]: trying URL 'https://cran.rstudio.com/src/contrib/mice_3.13.0.tar.gz'

R[write to console]: Content type 'application/x-gzip'
R[write to console]:  length 574057 bytes (560 KB)

R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write 

In [12]:
%%R
#sprecyzowanie metod imputacji
meth[c("Cholesterol")]="norm" 
meth[c("Smoking")]="logreg" 
meth[c("Education")]="polyreg"

In [13]:
%%R
set.seed(103)
imputed = mice(dat, method=meth, predictorMatrix=predM, m=5)


 iter imp variable
  1   1  Cholesterol  BMI  Smoking  Education
  1   2  Cholesterol  BMI  Smoking  Education
  1   3  Cholesterol  BMI  Smoking  Education
  1   4  Cholesterol  BMI  Smoking  Education
  1   5  Cholesterol  BMI  Smoking  Education
  2   1  Cholesterol  BMI  Smoking  Education
  2   2  Cholesterol  BMI  Smoking  Education
  2   3  Cholesterol  BMI  Smoking  Education
  2   4  Cholesterol  BMI  Smoking  Education
  2   5  Cholesterol  BMI  Smoking  Education
  3   1  Cholesterol  BMI  Smoking  Education
  3   2  Cholesterol  BMI  Smoking  Education
  3   3  Cholesterol  BMI  Smoking  Education
  3   4  Cholesterol  BMI  Smoking  Education
  3   5  Cholesterol  BMI  Smoking  Education
  4   1  Cholesterol  BMI  Smoking  Education
  4   2  Cholesterol  BMI  Smoking  Education
  4   3  Cholesterol  BMI  Smoking  Education
  4   4  Cholesterol  BMI  Smoking  Education
  4   5  Cholesterol  BMI  Smoking  Education
  5   1  Cholesterol  BMI  Smoking  Education
  5   2  Chole

In [14]:
%%R
imputed <- complete(imputed)
sapply(imputed, function(x) sum(is.na(x)))

        Age      Gender Cholesterol  SystolicBP         BMI     Smoking 
          5           0           0           0           0           0 
  Education 
          1 


### Sprawdzenie poprawności imputacji


In [37]:
%%R
summary(original)

      Age           Gender           Cholesterol      SystolicBP   
 Min.   :33.70   Length:250         Min.   :170.3   Min.   :117.5  
 1st Qu.:51.50   Class :character   1st Qu.:207.9   1st Qu.:138.6  
 Median :56.90   Mode  :character   Median :222.7   Median :145.4  
 Mean   :57.63                      Mean   :224.6   Mean   :145.5  
 3rd Qu.:63.20                      3rd Qu.:241.6   3rd Qu.:153.2  
 Max.   :79.30                      Max.   :303.0   Max.   :175.4  
 NA's   :5                          NA's   :20                     
      BMI        Smoking     Education 
 Min.   :13.80   No  :164   High  :99  
 1st Qu.:23.70   Yes : 66   Low   :64  
 Median :26.60   NA's: 20   Medium:67  
 Mean   :26.74              NA's  :20  
 3rd Qu.:29.50                         
 Max.   :38.70                         
 NA's   :5                             


In [32]:
%%R
summary(imputed)

      Age           Gender           Cholesterol      SystolicBP   
 Min.   :33.70   Length:250         Min.   :170.3   Min.   :117.5  
 1st Qu.:51.50   Class :character   1st Qu.:208.7   1st Qu.:138.6  
 Median :56.90   Mode  :character   Median :223.6   Median :145.4  
 Mean   :57.63                      Mean   :225.1   Mean   :145.5  
 3rd Qu.:63.20                      3rd Qu.:242.1   3rd Qu.:153.2  
 Max.   :79.30                      Max.   :303.0   Max.   :175.4  
 NA's   :5                                                         
      BMI        Smoking    Education  
 Min.   :13.80   No :179   High  :110  
 1st Qu.:23.93   Yes: 71   Low   : 69  
 Median :26.65             Medium: 70  
 Mean   :26.79             NA's  :  1  
 3rd Qu.:29.50                         
 Max.   :38.70                         
                                       


Jak widać powyżej, po imputacji dla zmiennej Cholesterol średnia wynosi 225.1 a w oryginalnych danych, w których brakowało 20 danych, wynosiła 224.6.

W przypadku zmiennej Education, pozostał 1 brak danych po imputacji. Po imputacji odpowiedź "High" stanowi 44% znanych odpowiedzi, natomiast przed stanowiła 43%. W przypadku odpowiedzi "Medium" to po imputacji 20% a przed 29%. W przypadku "Low" to 28% po imputacji jak i przed nią.

## Przykład 2 

# Źródła:


*   https://www.ajs.or.at/index.php/ajs/article/view/vol33%2C%20no1%262%20-%209
*   https://data.library.virginia.edu/getting-started-with-multiple-imputation-in-r/
*   https://www.statisticssolutions.com/dissertation-resources/multiple-imputation-for-missing-data/
*   https://www.analyticsvidhya.com/blog/2016/03/tutorial-powerful-packages-imputing-missing-values/
*   https://datascienceplus.com/handling-missing-data-with-mice-package-a-simple-approach/

